[SPARK-31801][API][SHUFFLE] Register map output metadata

mccheah opened the pull request:
### What changes were proposed in this pull request?

Adds a `ShuffleOutputTracker` API that can be used for managing shuffle metadata on the driver. Accepts map output metadata returned by the map output writers.

Requires https://github.com/apache/spark/pull/28616.

### Why are the changes needed?

Part of the design as discussed in [this document](https://docs.google.com/document/d/1Aj6IyMsbS2sdIfHxLvIbHUNjHIWHTabfknIPoxOrTjk/edit#), and part of the wider effort of SPARK-25299.

### Does this PR introduce _any_ user-facing change?

Enables additional APIs for the shuffle storage plugin tree. Usage will become more apparent when the read side of the shuffle plugin tree is introduced.

### How was this patch tested?

We've added a mock implementation of the shuffle plugin tree here, to prove that a Spark job using a different implementation of the plugin can use all of the plugin points for an alternative shuffle data storage solution. But we don't include it here, in order to minimize the diff and the code to review in this specific patch. See https://github.com/apache/spark/pull/28902.

mccheah committed:
Register map output metadata upon committing shuffle map tasks.

mccheah committed:
Add a mock version for tests.

SparkQA commented:
**[Test build #123022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123022/testReport)** for PR 28618 at commit [`98821b9`](https://github.com/apache/spark/commit/98821b91a0273300ffb40c197628f1abf69fbafb).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/27665/
Test PASSed.

SparkQA commented:
**[Test build #123022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123022/testReport)** for PR 28618 at commit [`98821b9`](https://github.com/apache/spark/commit/98821b91a0273300ffb40c197628f1abf69fbafb).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123022/
Test FAILed.

shadowinlife commented on core/src/main/java/org/apache/spark/shuffle/api/ShuffleMapOutputWriter.java:
```diff
@@ -63,7 +64,7 @@
    * The returned array should contain, for each partition from (0) to (numPartitions - 1), the
    * number of bytes written by the partition writer for that partition id.
    */
```
Suggest change the comments of this code

shadowinlife commented on core/src/main/java/org/apache/spark/shuffle/api/ShuffleMapOutputWriter.java:
```diff
@@ -63,7 +64,7 @@
    * number of bytes written by the partition writer for that partition id.
    */
-  long[] commitAllPartitions() throws IOException;
+  MapOutputCommitMessage commitAllPartitions() throws IOException;
```
Suggest change the relate comments of this code, the return is an object instead of array

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28095/
Test PASSed.

mccheah committed:
Update manifests

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28125/
Test PASSed.

SparkQA commented:
**[Test build #123502 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123502/testReport)** for PR 28618 at commit [`25e98e7`](https://github.com/apache/spark/commit/25e98e7a0c3a44b4b07c3ef2f73b137fc1ecc384).

SparkQA commented:
**[Test build #123504 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123504/testReport)** for PR 28618 at commit [`51df151`](https://github.com/apache/spark/commit/51df1513cbca534b0f3e577f720c5cae9f44a870).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28127/
Test PASSed.

mccheah committed:
Add a test for specifically checking map output registration and backup

SparkQA commented:
**[Test build #123507 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123507/testReport)** for PR 28618 at commit [`e7c9988`](https://github.com/apache/spark/commit/e7c998814df365c726d6615da4b7a14b2ba2167c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28131/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123502/
Test FAILed.

SparkQA commented:
**[Test build #123504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123504/testReport)** for PR 28618 at commit [`51df151`](https://github.com/apache/spark/commit/51df1513cbca534b0f3e577f720c5cae9f44a870).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123504/
Test FAILed.

SparkQA commented:
**[Test build #123507 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123507/testReport)** for PR 28618 at commit [`e7c9988`](https://github.com/apache/spark/commit/e7c998814df365c726d6615da4b7a14b2ba2167c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123507/
Test FAILed.

attilapiros commented on core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:
```diff
@@ -130,9 +132,13 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
-        mapStatus = MapStatus$.MODULE$.apply(
-          blockManager.shuffleServerId(), partitionLengths, mapId);
+        mapOutputCommitMessage = mapOutputWriter.commitAllPartitions();
+        taskResult = new MapTaskResult(
```
As these lines are repeating you could extract them into a new `def`, like:

```scala
  protected void setTaskResult(MapOutputCommitMessage mapOutputCommitMessage) {
    taskResult = new MapTaskResult(
        MapStatus$.MODULE$.apply(
            blockManager.shuffleServerId(),
            mapOutputCommitMessage.getPartitionLengths(),
            mapId),
        OptionConverters.toScala(mapOutputCommitMessage.getMapOutputMetadata()));
  }
```

With the help of this new `def` and Mockito's spy you can even get rid of the  storing the `mapOutputCommitMessage` for testing purposes only but it has a price (this class cannot be final) for details you can check:
https://github.com/attilapiros/spark/commit/f4578a31beccca58e8b55120847db014a30159d6

mccheah replied:
Ack - didn't address this in my latest patch but will get around to this

attilapiros commented on core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java:
```diff
@@ -273,19 +280,24 @@ void forceSorterToSpill() throws IOException {
+        Optional<MapOutputMetadata> maybeMetadata =
+            maybeSingleFileWriter.get().transferMapSpillFile(
+                spills[0].file, spills[0].partitionLengths);
+        mapOutputCommitMessage = maybeMetadata.map(
+            metadata -> MapOutputCommitMessage.of(spills[0].partitionLengths, metadata))
+            .orElse(MapOutputCommitMessage.of(spills[0].partitionLengths));
```
I cannot se why `transferMapSpillFile` cannot return a `MapOutputCommitMessage` that would simply this part:
https://github.com/attilapiros/spark/commit/289050e4309b024146867b7b32974ab4746eb570

mccheah replied:
Hm, I think this was originally designed this way because we didn't want the single spill writer to set a list of partition lengths that was different from what was passed into the writer's transfer function. But, maybe we can wrap this with a preconditions check to ensure that the state remains consistent, and that's good enough along with Javadoc.

attilapiros commented:
I was thinking a lot on `ShuffleDriverComponents` and I have an idea how to improve it. 

The problem I believe this class tries to fulfil two very separate roles: be a **builder** and the **result of the building** in the same time. 

This is why we need this kind of check:
https://github.com/apache/spark/blob/289050e4309b024146867b7b32974ab4746eb570/core/src/main/java/org/apache/spark/shuffle/sort/io/LocalDiskShuffleDriverComponents.java#L47-L49 

If the building is cleanly separated from the result of the building then we can be sure the prerequisites are fulfilled before. 

I would change it by transforming it to be the result of the building in the following way:
- the `initializeApplication` (I mean the process of the initialisation and not the returned Map) should be part of the building. The documentation of the `ShuffleDataIO#driver` method can be extended by mentioning this is the right place to initialize.
- the `ShuffleDriverComponents` could have a new method which gives back the "additional SparkConf settings necessary for initializing the executor components" we can call it like `additonalExecutorConfigs`. This new method would replace the old `initializeApplication`

One more idea / question:
- I do not see why the `ShuffleOutputTracker` is optional. Either we or the API user can provide an implementation where the methods are empty this way the API a bit simpler.

@mccheah what do you think?

SparkQA commented:
**[Test build #124261 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124261/testReport)** for PR 28618 at commit [`e7c9988`](https://github.com/apache/spark/commit/e7c998814df365c726d6615da4b7a14b2ba2167c).

SparkQA commented:
**[Test build #124261 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124261/testReport)** for PR 28618 at commit [`e7c9988`](https://github.com/apache/spark/commit/e7c998814df365c726d6615da4b7a14b2ba2167c).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124261/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #124307 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124307/testReport)** for PR 28618 at commit [`e7c9988`](https://github.com/apache/spark/commit/e7c998814df365c726d6615da4b7a14b2ba2167c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28926/
Test PASSed.

SparkQA commented:
**[Test build #124307 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124307/testReport)** for PR 28618 at commit [`e7c9988`](https://github.com/apache/spark/commit/e7c998814df365c726d6615da4b7a14b2ba2167c).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124307/
Test FAILed.

mccheah commented:
> I was thinking a lot on `ShuffleDriverComponents` and I have an idea how to improve it.
> 
> The problem I believe this class tries to fulfil two very separate roles: be a **builder** and the **result of the building** in the same time.
> 
> This is why we need this kind of check:
> https://github.com/apache/spark/blob/289050e4309b024146867b7b32974ab4746eb570/core/src/main/java/org/apache/spark/shuffle/sort/io/LocalDiskShuffleDriverComponents.java#L47-L49
> 
> If the building is cleanly separated from the result of the building then we can be sure the prerequisites are fulfilled before.
> 
> I would change it by transforming it to be the result of the building in the following way:
> 
> * the `initializeApplication` (I mean the process of the initialisation and not the returned Map) should be part of the building. The documentation of the `ShuffleDataIO#driver` method can be extended by mentioning this is the right place to initialize.
> * the `ShuffleDriverComponents` could have a new method which gives back the "additional SparkConf settings necessary for initializing the executor components" we can call it like `additonalExecutorConfigs`. This new method would replace the old `initializeApplication`
> 
> One more idea / question:
> 
> * I do not see why the `ShuffleOutputTracker` is optional. Either we or the API user can provide an implementation where the methods are empty this way the API a bit simpler.
> 
> @mccheah what do you think?

You bring up a good point. I can adjust the PR accordingly. It does seem like the components does both an initialization and a runtime mode, and it would be more ideal to separate the two. Thanks for critically thinking about this!

mccheah committed:
Return map output commit message from single spill writer

mccheah commented:
Also I think it makes sense for the executor side and the driver side to be mirrored.

mccheah committed:
Address comments and fix build

mccheah committed:
Temporarily remove tests

mccheah commented:
I rebased on master in my latest patch. I also addressed your comments @attilapiros. Thanks for the feedback!

The diff was growing extremely large (> 1000 lines), so I removed all the tests for now. I'm going to open a separate patch with tests for this.

SparkQA commented:
**[Test build #124385 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124385/testReport)** for PR 28618 at commit [`dc8d15c`](https://github.com/apache/spark/commit/dc8d15c4078774dce031096b418695b51fb100ee).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/29004/
Test PASSed.

SparkQA commented:
**[Test build #124385 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/124385/testReport)** for PR 28618 at commit [`dc8d15c`](https://github.com/apache/spark/commit/dc8d15c4078774dce031096b418695b51fb100ee).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/124385/
Test FAILed.

attilapiros commented:
In `BaseReceivedBlockHandlerSuite` one parameter (a `ShuffleOutputTracker`) is missing but as the test does not do any shuffle we can pass a `null` as value (or a `NoOpShuffleOutputTracker`). With `null`:

```diff
index 0976494b6d..5f6a0f164f 100644
--- a/streaming/src/test/scala/org/apache/spark/streaming/ReceivedBlockHandlerSuite.scala
+++ b/streaming/src/test/scala/org/apache/spark/streaming/ReceivedBlockHandlerSuite.scala
@@ -70,7 +70,7 @@ abstract class BaseReceivedBlockHandlerSuite(enableEncryption: Boolean)
   val streamId = 1
   val securityMgr = new SecurityManager(conf, encryptionKey)
   val broadcastManager = new BroadcastManager(true, conf, securityMgr)
-  val mapOutputTracker = new MapOutputTrackerMaster(conf, broadcastManager, true)
+  val mapOutputTracker = new MapOutputTrackerMaster(conf, null, broadcastManager, true)
   val shuffleManager = new SortShuffleManager(conf)
   val serializer = new KryoSerializer(conf)
   var serializerManager = new SerializerManager(serializer, conf, encryptionKey)
``` 

```
 [info] ReceivedBlockHandlerSuite:
  [info] - BlockManagerBasedBlockHandler - store blocks (315 milliseconds)
  [info] - BlockManagerBasedBlockHandler - handle errors in storing block (5 milliseconds)
  [info] - WriteAheadLogBasedBlockHandler - store blocks (216 milliseconds)
  [info] - WriteAheadLogBasedBlockHandler - handle errors in storing block (13 milliseconds)
  [info] - WriteAheadLogBasedBlockHandler - clean old blocks (53 milliseconds)
  [info] - Test Block - count messages (121 milliseconds)
  [info] - Test Block - isFullyConsumed (24 milliseconds)
  [info] ReceivedBlockHandlerWithEncryptionSuite:
  [info] - BlockManagerBasedBlockHandler - store blocks (27 milliseconds)
  [info] - BlockManagerBasedBlockHandler - handle errors in storing block (2 milliseconds)
  [info] - WriteAheadLogBasedBlockHandler - store blocks (89 milliseconds)
  [info] - WriteAheadLogBasedBlockHandler - handle errors in storing block (8 milliseconds)
  [info] - WriteAheadLogBasedBlockHandler - clean old blocks (16 milliseconds)
  [info] - Test Block - count messages (71 milliseconds)
  [info] - Test Block - isFullyConsumed (14 milliseconds)
  [info] ScalaTest
  [info] Run completed in 4 seconds, 30 milliseconds.
  [info] Total number of tests run: 14
  [info] Suites: completed 2, aborted 0
  [info] Tests: succeeded 14, failed 0, canceled 0, ignored 0, pending 0
  [info] All tests passed.
```

attilapiros commented on core/src/main/scala/org/apache/spark/shuffle/MemoizingShuffleDataIO.scala:
```diff
@@ -0,0 +1,51 @@
+ * and then reuses them throughout the lifetime of the SparkEnv.
+ */
+class MemoizingShuffleDataIO(delegate: ShuffleDataIO) {
+  private var _driver: ShuffleDriverComponents = _
+  private var _executor: ShuffleExecutorComponents = _
+
+  def getOrCreateDriverComponents(): ShuffleDriverComponents = synchronized {
+    if (_driver == null) {
+      _driver = delegate.initializeShuffleDriverComponents()
+    }
+    _driver
+  }
+
+  def getOrCreateExecutorComponents(
+      appId: String,
+      execId: String,
+      extraConfigs: Map[String, String]): ShuffleExecutorComponents = synchronized {
+    if (_executor == null) {
+      _executor = delegate.initializeShuffleExecutorComponents(appId, execId, extraConfigs.asJava)
+    }
+    _executor
+  }
```
I think we can improve the methods of this class. Especially with `getOrCreateExecutorComponents` as reading that I have a feeling it is easy to use it badly by calling with two different values for any of its params right after each other as always the object generated for the first call will be returned.

But first `getOrCreateDriverComponents` that is the easier as here we have no input params at all so this could be replaced with a simple `lazy val` (it is thread-safe for a long time ago: https://github.com/scala/bug/issues/3007).

And the same is true for `getOrCreateExecutorComponents ` as all its parameters are basically coming from a SparkEnv instance.

So my idea is:
```suggestion

  lazy val driverComponents = delegate.initializeShuffleDriverComponents()

  lazy val executorComponents = {
    val env = SparkEnv.get
    delegate.initializeShuffleExecutorComponents(
      env.conf.getAppId,
      env.executorId,
      env.conf.getAllWithPrefix(ShuffleDataIOUtils.SHUFFLE_SPARK_CONF_PREFIX).toMap.asJava)
  }
```
I still have to test it. What is your opinion @mccheah ?

mccheah replied:
I certainly gave `lazy val` a consideration. I'm not entirely familiar with the most modern Scala conventions - if `lazy` is preferred over explicit initialization methods in general, then I'm ok with the above recommendation.

attilapiros replied:
I think transforming `getOrCreateExecutorComponents` into method without any argument is already big win.
I can let go the `lazy val` it is just implementation details.

mccheah replied:
LGTM, I'll make those changes. Thanks!

mccheah committed:
Address comments

mccheah committed:
Merge remote-tracking branch 'origin/master' into register-map-output-metadata

mccheah commented:
Took a pass through this to address @attilapiros's latest comments and also to merge with the upstream master branch. Is there any more feedback for this PR, or this feature as a whole, the API design, etc?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31454/
Test PASSed.

mccheah committed:
Fix Java linting errors

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31459/
Test PASSed.

SparkQA commented:
**[Test build #126831 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126831/testReport)** for PR 28618 at commit [`edd5c05`](https://github.com/apache/spark/commit/edd5c0541fe14a412d073d9ed1f7019d294c3425).

SparkQA commented:
**[Test build #126836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126836/testReport)** for PR 28618 at commit [`7baa3d2`](https://github.com/apache/spark/commit/7baa3d27ce06f136b5edaabac7177669346ba38d).

SparkQA commented:
**[Test build #126831 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126831/testReport)** for PR 28618 at commit [`edd5c05`](https://github.com/apache/spark/commit/edd5c0541fe14a412d073d9ed1f7019d294c3425).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126831/
Test FAILed.

SparkQA commented:
**[Test build #126836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126836/testReport)** for PR 28618 at commit [`7baa3d2`](https://github.com/apache/spark/commit/7baa3d27ce06f136b5edaabac7177669346ba38d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126836/
Test FAILed.

mccheah committed:
Fix infinite loop.

Only invoke shuffle output tracker once per unregister shuffle attempt.

SparkQA commented:
**[Test build #126894 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126894/testReport)** for PR 28618 at commit [`cce67ab`](https://github.com/apache/spark/commit/cce67abf51f2cadf1cd36cd0971d8545fc9fcb65).

mccheah committed:
Fix checkstyle

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31519/
Test PASSed.

SparkQA commented:
**[Test build #126896 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126896/testReport)** for PR 28618 at commit [`409bebb`](https://github.com/apache/spark/commit/409bebbcab551efd0f29da590d22925cad30d465).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31521/
Test PASSed.

SparkQA commented:
**[Test build #126894 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126894/testReport)** for PR 28618 at commit [`cce67ab`](https://github.com/apache/spark/commit/cce67abf51f2cadf1cd36cd0971d8545fc9fcb65).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126894/
Test FAILed.

SparkQA commented:
**[Test build #126896 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126896/testReport)** for PR 28618 at commit [`409bebb`](https://github.com/apache/spark/commit/409bebbcab551efd0f29da590d22925cad30d465).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126896/
Test FAILed.

mccheah committed:
Fix test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31527/
Test PASSed.

SparkQA commented:
**[Test build #126904 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126904/testReport)** for PR 28618 at commit [`a6fabd2`](https://github.com/apache/spark/commit/a6fabd2c79e5ce7837b526e06c5d9e64b03117b1).

SparkQA commented:
**[Test build #126904 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126904/testReport)** for PR 28618 at commit [`a6fabd2`](https://github.com/apache/spark/commit/a6fabd2c79e5ce7837b526e06c5d9e64b03117b1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126904/
Test FAILed.

mccheah committed:
Invoke super.afterEach in ShuffleDriverComponentsSuite

Cause we need to stop the SC

SparkQA commented:
**[Test build #126998 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126998/testReport)** for PR 28618 at commit [`3c66353`](https://github.com/apache/spark/commit/3c6635344dcc6c25da6866ba9ea59c0f37f64c20).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31622/
Test PASSed.

SparkQA commented:
**[Test build #126998 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/126998/testReport)** for PR 28618 at commit [`3c66353`](https://github.com/apache/spark/commit/3c6635344dcc6c25da6866ba9ea59c0f37f64c20).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/126998/
Test FAILed.

mccheah committed:
Fix build

SparkQA commented:
**[Test build #127000 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127000/testReport)** for PR 28618 at commit [`b88d724`](https://github.com/apache/spark/commit/b88d7249ba8adcc0665e5f445b22d090e4741f38).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31624/
Test PASSed.

SparkQA commented:
**[Test build #127000 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127000/testReport)** for PR 28618 at commit [`b88d724`](https://github.com/apache/spark/commit/b88d7249ba8adcc0665e5f445b22d090e4741f38).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127000/
Test FAILed.

mccheah committed:
Merge remote-tracking branch 'origin/master' into register-map-output-metadata

SparkQA commented:
**[Test build #127012 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127012/testReport)** for PR 28618 at commit [`3505af8`](https://github.com/apache/spark/commit/3505af83f7713600dd43a0a489e8dbebf1c9e0a7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31636/
Test PASSed.

SparkQA commented:
**[Test build #127012 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127012/testReport)** for PR 28618 at commit [`3505af8`](https://github.com/apache/spark/commit/3505af83f7713600dd43a0a489e8dbebf1c9e0a7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class RegExpExtractAll(subject: Expression, regexp: Expression, idx: Expression)`
  * `    .doc(\"The name of a class that implements \" +`
  * `trait CachedBatch `
  * `trait CachedBatchSerializer extends Serializable `
  * `trait SimpleMetricsCachedBatch extends CachedBatch `
  * `abstract class SimpleMetricsCachedBatchSerializer extends CachedBatchSerializer with Logging `
  * `case class ColumnarToRowExec(child: SparkPlan) extends ColumnarToRowTransition with CodegenSupport `
  * `case class ApplyColumnarRulesAndInsertTransitions(`
  * `class ColumnStatisticsSchema(a: Attribute) extends Serializable `
  * `class PartitionStatistics(tableSchema: Seq[Attribute]) extends Serializable `
  * `case class DefaultCachedBatch(numRows: Int, buffers: Array[Array[Byte]], stats: InternalRow)`
  * `class DefaultCachedBatchSerializer extends SimpleMetricsCachedBatchSerializer `
  * `trait HashJoin extends BaseJoinExec with CodegenSupport `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127012/
Test PASSed.

mccheah commented:
Ok, I think the build passes according to Jenkins, though the tests run by Github actions appears to have been flaky. This is ready for review or for sign off.

tianczha commented on core/src/main/java/org/apache/spark/shuffle/sort/io/LocalDiskShuffleDataIO.java:
```diff
@@ -35,12 +37,13 @@ public LocalDiskShuffleDataIO(SparkConf sparkConf) {
 
   @Override
-  public ShuffleExecutorComponents executor() {
+  public ShuffleExecutorComponents initializeShuffleExecutorComponents(
```
Can we move `initializeShuffleDriverComponents` before `initializeShuffleExecutorComponents`? 
The extraConfigs will be only available after driver components is initialized, right?

tianczha commented on core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala:
```diff
@@ -20,7 +20,7 @@ package org.apache.spark.shuffle
 import org.apache.spark.internal.Logging
 import org.apache.spark.rdd.RDD
-import org.apache.spark.scheduler.MapStatus
+import org.apache.spark.scheduler.{MapStatus, MapTaskResult}
```
remove MapStatus?

mccheah committed:
Address comments.

SparkQA commented:
**[Test build #127295 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127295/testReport)** for PR 28618 at commit [`89bb528`](https://github.com/apache/spark/commit/89bb528d54949e5f3cac623bf01e253d39d68ced).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31914/
Test PASSed.

SparkQA commented:
**[Test build #127295 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127295/testReport)** for PR 28618 at commit [`89bb528`](https://github.com/apache/spark/commit/89bb528d54949e5f3cac623bf01e253d39d68ced).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127295/
Test PASSed.

mridulm commented:
+CC @otterc

mccheah commented:
Ping here - what would be the best way to get this merged?

mridulm commented:
@mccheah I will take a look at this later this week/early next week.
+CC @squito, @holdenk who reviewed the design doc.

mridulm commented on core/src/main/java/org/apache/spark/shuffle/api/metadata/ShuffleOutputTracker.java:
```diff
@@ -0,0 +1,56 @@
+   * Called when a new shuffle stage is going to be run.
+   */
+  void registerShuffle(int shuffleId);
+
```
A callback to signal completion (or failure) of shuffle for a `shuffleId` will help `ShuffleDriverComponents` coordinate with external services to commit (or cleanup) shuffle metadata/files.

mccheah replied:
I'm hesitant to make significant API changes given how late this is in the review stage - let's add it as a follow-up if there are concrete use cases that require it down the road.

mridulm replied:
Looking at the currently designed api, this is a missing gap.

mccheah replied:
Again - I don't want to add it here, since it would require further integration of such a new API call in the rest of the Spark codebase, which I am holding as out of scope for this patch.

mridulm replied:
This is as a gap in the proposed api - which limits the effectiveness to leverage it in for other shuffle usecases.
+CC @otterc PTAL if without this, shuffle changes can be made for your patches.

mccheah replied:
This API isn't complete as of this patch anyways, so I'd prefer that functionality to be deferred since the complexity being added here is already pretty significant. Can we file a follow up JIRA for it and go from there?

mridulm commented on core/src/main/java/org/apache/spark/shuffle/sort/io/LocalDiskShuffleDataIO.java:
```diff
@@ -35,12 +37,13 @@ public LocalDiskShuffleDataIO(SparkConf sparkConf) {
-    return new LocalDiskShuffleDriverComponents();
+  public ShuffleExecutorComponents initializeShuffleExecutorComponents(
+      String appId, String execId, Map<String, String> extraConfigs) {
+    return new LocalDiskShuffleExecutorComponents(sparkConf);
```
nit: ignoring `extraConfigs` - assert it is empty ?

mccheah replied:
Don't think it's particularly important that it's empty and ignored - maybe just add a warning log.

mridulm commented on core/src/main/java/org/apache/spark/shuffle/sort/io/LocalDiskShuffleExecutorComponents.java:
```diff
@@ -17,69 +17,64 @@
   private final SparkConf sparkConf;
-  private BlockManager blockManager;
-  private IndexShuffleBlockResolver blockResolver;
+  private final Supplier<IndexShuffleBlockResolver> blockResolver;
```
Here and elsewhere, why rely on `Supplier` when output is memoized ?
Replace with instance instead ?

mccheah replied:
I believe this is a supplier because at the time the executor components are initialized, the entire SparkEnv instance is not initialized yet, so we want the block resolver to be fetched from the SparkEnv lazily.

mridulm replied:
Can you elaborate on this ? An example where `SparkEnv` was not initialized but executor shuffle component was ?
The usages I saw all had `SparkEnv` already available, but I could have missed some flow.

mridulm replied:
The flow I see (ignoring tests) is :
`MemoizingShuffleDataIO._executor` -> `LocalDiskShuffleDataIO.initializeShuffleExecutorComponents` -> `LocalDiskShuffleExecutorComponents ctor`

 `_executor` is a `lazy val`, but its construction requires `SparkEnv` to have been set (and so initialized).
For this path, we will have block manager from `SparkEnv` available to create `IndexShuffleBlockResolver` in `LocalDiskShuffleExecutorComponents`.

Is there any other path or flow which requires this to be lazy ?

mccheah replied:
Hm, I think that now that the laziness is handled at the level above (MemoizingShuffleDataIO) it should be fine to make these non-lazy then. I'll try that and push and check the build after. I recall there being something weird with local mode, but that might have been a state that isn't possible now with the new memoizing layer above.

mridulm commented on core/src/main/scala/org/apache/spark/ContextCleaner.scala:
```diff
@@ -222,15 +220,10 @@ private[spark] class ContextCleaner(
-      } else {
-        logDebug("Asked to cleanup non-existent shuffle (maybe it was already removed)")
-      }
+      logDebug("Cleaning shuffle " + shuffleId)
```
Why was the containsShuffle check removed ?

mridulm commented on core/src/main/java/org/apache/spark/shuffle/api/metadata/ShuffleOutputTracker.java:
```diff
@@ -0,0 +1,56 @@
+ * {@link MapOutputMetadata} instances provided from the plugin tree's implementation of
+ * {@link org.apache.spark.shuffle.api.ShuffleMapOutputWriter} are sent to
+ * <p>
+ * Implementations MUST be thread-safe. Spark will invoke methods in this module in parallel.
```
Add a note about what locking semantics are when the methods are invoked ?

mccheah replied:
I'm not sure what can be added. Can you give an example of guidance we could give to developers?

mridulm replied:
As an example, `registerShuffle` is invoked with write lock held, while `unregisterShuffle` is invoked without any lock.
As a public spi we are exposing, the locking semantics needs to be elaborated; so that users can evaluate what can be done safely/without contention, and what needs additional protection (for example, `updateMapOutput` can be performed without contention from `registerShuffle`)

Additionally, it also informs users about how to design their implementations. For example, if an implementation makes rpc calls (or other blocking calls) from `registerMapOutput` , there could be nontrivial side effects in driver.

(I just picked some api calls to illustrate, we need to elaborate for each).

mccheah replied:
Is it sufficient to simply say that all APIs need to have exclusive access from each other, at least locking on the shuffle id level?

mridulm replied:
`unregisterShuffle` currently does not lock ?

mccheah replied:
Again, that's up to the implementation to handle, no?

mridulm replied:
I was not referring to how implementations manage their state/locking (if they have additional state/coordination required, you are right, it has to be handled there - state managed across shuffles as an example for some custom impl).

I was referring to what guarantees/expectations that implementations have from spark (`MapOutputTracker` currently) when these methods are invoked. We have to document this, so that implementation can be both aware of the MT-safety guarantees we provide (and at what granularity) - so that they can make reasonable assumptions about what can/cant be done in an implementation.

For example, as mentioned above, `registerShuffle` is invoked only with write lock is held for a particular shuffle. While `unregisterShuffle` does not have that guarantee.

mccheah replied:
The whole point of just saying that implementations must be thread-safe is that implementations cannot make _any_ guarantees about how the caller will invoke this API. Even if we can say that the current implementation of `MapOutputTracker` would guarantee locking on per-shuffle-id, for example, that isn't an API guarantee we can realistically make since `MapOutputTracker` is a module internal to Spark.

I think saying that there is no guarantee about concurrent access to all methods suffices. But then that basically means "the implementation should be thread-safe". Is that reasonable, or do we need to be more granular than that?

mridulm reviewed (commented):
Took one pass through the PR - thanks for working on this @mccheah - very nice surgical change.
@tgravescs Can you also take a look ? Given I have not touched this part in a while, want to make sure there is more review on the change. Thx

mridulm commented:
+CC @tgravescs 
This is part of the changes currently being reviewed, which could impact push based shuffle.

tgravescs commented:
thanks @mridulm I need to catch up on the docs and I'll take a look.

mccheah commented on core/src/main/java/org/apache/spark/shuffle/api/metadata/ShuffleOutputTracker.java:
```diff
@@ -0,0 +1,56 @@
+package org.apache.spark.shuffle.api.metadata;
+
+/**
+ * :: Private ::
```
I believe this is being marked as private until we have the entire API completed and available to implement end to end. Right now the API is not really usable in any meaningful way.

mridulm commented:
+CC @jiangxb1987

mccheah commented on core/src/main/scala/org/apache/spark/shuffle/MemoizingShuffleDataIO.scala:
```diff
@@ -0,0 +1,45 @@
+ * Used to ensure the SparkEnv only instantiates the given components once lazily
+ * and then reuses them throughout the lifetime of the SparkEnv.
+ */
+class MemoizingShuffleDataIO(delegate: ShuffleDataIO) {
```
It's because everything has to be instantiated lazily, basically.

mccheah committed:
Address comments

mccheah committed:
Merge remote-tracking branch 'origin/master' into register-map-output-metadata

SparkQA commented:
**[Test build #128474 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128474/testReport)** for PR 28618 at commit [`2b5108f`](https://github.com/apache/spark/commit/2b5108f349e1d42c89e1e13c844d292167948b41).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33098/
Test PASSed.

mccheah commented:
@tgravescs sorry for the delay - addressed all the comments above.

SparkQA commented:
**[Test build #128474 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128474/testReport)** for PR 28618 at commit [`2b5108f`](https://github.com/apache/spark/commit/2b5108f349e1d42c89e1e13c844d292167948b41).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128474/
Test PASSed.

mridulm commented:
Btw, noticed that `updateMapOutput` was added as part of executor decommission support. We will need to hook that into `ShuffleOutputTracker` as well.

+CC @holdenk

otterc commented:
I looked at the changes proposed here so that we can use the interfaces here for Push-based shuffle ([SPIP](https://docs.google.com/document/d/1mYzKVZllA5Flw8AtoX7JUcXBOnNIDADWRbJ7GI6Y71Q/edit?ts=5f5d1718) and [code](https://github.com/linkedin/spark/tree/magnet-upstream)). 
In Push-based shuffle, we have introduced [merge-statuses](https://github.com/linkedin/spark/blob/magnet-upstream/core/src/main/scala/org/apache/spark/scheduler/MergeStatus.scala) which represent all the map outputs that were merged into a larger block. These statuses are collected by the driver from the Shuffle Services. 

I think we will be able to use the current `ShuffleOutputTracker` API. The implementation of this API could have the triggers for finalizing the shuffle merge.

I still have to wrap my head around how we can model `mergeStatus` as part of `MapOutputMetadata`. Multiple`mapStatus`'s would point to a single `mergeStatus`, so this would introduce some complexity.

We may need to evolve them to fit the push-based shuffle use case. As long as we are open to potentially making some backward incompatible changes, these APIs look good to me for now.

mccheah commented:
> Btw, noticed that updateMapOutput was added as part of executor decommission support. We will need to hook that into ShuffleOutputTracker as well.

@mridulm Can we also do this as follow-up? The main thing is, this patch is already at ~900 lines changed total (+s and -s combined) and I really don't want to increase the scope of this. This patch has already stalled from merging for awhile and I'd rather get something completed and have follow-up tasks than try to make the entire feature perfect in a single patch.

Can we add follow-up JIRA tasks that followed from the comments for purely additive changes to the API, and move forward with the scope of this patch as-is?

mridulm commented:
Sure @mccheah we can do that in follow up work to keep things more bite sized.
I was making sure we dont miss out of it, and tagging holden on this.

attilapiros reviewed (approved):
LGTM (just a few nits)

holdenk commented:
Thanks for working on this :)

tgravescs commented:
sorry for my delay on getting back to this, could you up merge to latest?

attilapiros committed:
Merge branch 'master' into pr/28618

attilapiros committed:
apply Attila's review comments

attilapiros committed:
remove unused import

mccheah committed:
Merge pull request #15 from attilapiros/updated_28618

Updating PR 28618 with master and applying my comments

SparkQA commented:
**[Test build #129906 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129906/testReport)** for PR 28618 at commit [`a6d974c`](https://github.com/apache/spark/commit/a6d974cd5d673a0f3da9910fd46e95ca71ce5543).

SparkQA commented:
**[Test build #129906 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129906/testReport)** for PR 28618 at commit [`a6d974c`](https://github.com/apache/spark/commit/a6d974cd5d673a0f3da9910fd46e95ca71ce5543).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129906/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34512/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34512/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34512/
Test PASSed.

attilapiros committed:
fix mima

mccheah committed:
Merge pull request #16 from attilapiros/updated_28618

fix mima

SparkQA commented:
**[Test build #129950 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129950/testReport)** for PR 28618 at commit [`f69cba7`](https://github.com/apache/spark/commit/f69cba7d0632e7b260e38a0aa16cf560bd0d94cc).

SparkQA commented:
**[Test build #129950 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129950/testReport)** for PR 28618 at commit [`f69cba7`](https://github.com/apache/spark/commit/f69cba7d0632e7b260e38a0aa16cf560bd0d94cc).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129950/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34555/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34555/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34555/
Test PASSed.

holdenk commented:
Jenkins retest this please.

SparkQA commented:
**[Test build #130236 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130236/testReport)** for PR 28618 at commit [`f69cba7`](https://github.com/apache/spark/commit/f69cba7d0632e7b260e38a0aa16cf560bd0d94cc).

SparkQA commented:
**[Test build #130236 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/130236/testReport)** for PR 28618 at commit [`f69cba7`](https://github.com/apache/spark/commit/f69cba7d0632e7b260e38a0aa16cf560bd0d94cc).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/130236/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34836/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34836/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34836/
Test PASSed.

attilapiros commented:
For helping @mccheah who is busy with other projects I cloned this PR as #30763.
I will keep that up-to-date with master and react to review comments.
So if this fine for you please continue the review discussions there.
